gpu/preconditioner/block_jacobi_kernels.cu

```diff
@@ -166,6 +166,143 @@ GKO_ENABLE_IMPLEMENTATION_SELECTION(select_apply, apply);
 using compiled_kernels = syn::compile_int_list<1, 13, 16, 32>;
 
 
+template <typename IndexType>
+__global__ void compare_adjacent_rows(size_type num_rows, int32 max_block_size,
+                                      const IndexType *__restrict__ row_ptrs,
+                                      const IndexType *__restrict__ col_idx,
+                                      bool *matching_prev_row)
+{
+    const auto global_tid = blockDim.x * blockIdx.x + threadIdx.x;
+    const auto local_tid = threadIdx.x % cuda_config::warp_size;
+    const auto warp_id = global_tid / cuda_config::warp_size;
+
+    if (warp_id >= num_rows - 1) {
+        return;
+    }
+
+    const auto prev_row_start = row_ptrs[warp_id];
+    const auto curr_row_start = row_ptrs[warp_id + 1];
+    const auto next_row_start = row_ptrs[warp_id + 2];
+
+    const auto nz_this_row = next_row_start - curr_row_start;
+    const auto nz_prev_row = curr_row_start - prev_row_start;
+
+    if (nz_this_row != nz_prev_row) {
+        matching_prev_row[warp_id] = false;
+        return;
+    }
+    size_type steps = ceildiv(nz_this_row, cuda_config::warp_size);
+    for (size_type i = 0; i < steps; i++) {
+        auto j = local_tid + i * cuda_config::warp_size;
+        auto prev_col = (prev_row_start + j < curr_row_start)
+                            ? col_idx[prev_row_start + j]
+                            : 0;
+        auto this_col = (prev_row_start + j < curr_row_start)
+                            ? col_idx[curr_row_start + j]
+                            : 0;
+        if (warp::any(prev_col != this_col)) {
+            matching_prev_row[warp_id] = false;
+            return;
+        }
+    }
+    matching_prev_row[warp_id] = true;
+}
+
+
+template <typename IndexType>
+__global__ void generate_natural_block_pointer(
+    size_type num_rows, int32 max_block_size,
+    const bool *__restrict__ matching_prev_row, IndexType *block_ptrs,
+    size_type *num_blocks_arr)
```

Comment:
`__restrict__` again on all pointers - in case you're wondering why I'm constantly complaining about these restricts, it's because that tells additional information to the compiler (that there's no aliasing), which it can use to better optimize the code.

(BTW, this was one of the questions in my NVIDIA interview, so they definitely think it's important.)